pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go

```diff
@@ -0,0 +1,186 @@
+	}, nil
+}
+
+func (r *REST) New() runtime.Object {
+	return &oauthv1.PersonalAccessToken{}
+}
+
+func (r *REST) NewList() runtime.Object {
+	return &oauthv1.PersonalAccessTokenList{}
+}
+
+func (r *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListGenericObj, err := r.accessTokenStorage.List(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenList := tokenListGenericObj.(*oauthapi.OAuthAccessTokenList)
+	if tokenList == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
+	}
+
+	return oauthAccessTokenListToUserOAuthAccessTokenList(tokenList), nil
+}
+
+func (r *REST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenericObj, err := r.accessTokenStorage.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	token := tokenGenericObj.(*oauthapi.OAuthAccessToken)
+	if token == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if token.UserName != ctxUserName { // don't reveal other people tokens' hashes
+		return nil, errors.NewNotFound(oauth.Resource("personaltoken"), name)
+	}
+
+	return &oauthapi.PersonalAccessToken{OAuthAccessToken: *token}, nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("context is nil")
+	}
+
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher.ResultChan())
+	go watcher.Watch()
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if _, err := r.Get(ctx, name, &metav1.GetOptions{}); err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
```

Comment:
here is a small risk for a race: you have to pass the uid of the returned object in the options below, and probably also do some translation of GetOptions to DeleteOptions above.